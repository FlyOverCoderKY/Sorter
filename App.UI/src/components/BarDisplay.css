.bar-display {
  width: 100%;
  height: 100%;
  min-height: 400px;
  display: flex;
  flex-direction: column;
  position: relative;
}

.bar-container {
  flex: 1;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  padding: 24px;
  position: relative;
  overflow: hidden;
  background: var(--color-panel-background);
  border: 1px solid var(--color-border);
  border-radius: 8px;
}

.bar {
  border-radius: 2px 2px 0 0;
  min-height: 4px;
  position: relative;
  cursor: pointer;
  box-shadow: 0 1px 3px var(--color-shadow);
  /* Remove all transitions to prevent flickering during sorting */
  transition: none;
}

.bar:hover {
  transform: scaleY(1.05);
  box-shadow: 0 2px 6px var(--color-shadow-strong);
  /* Add transition only on hover */
  transition: transform 0.2s ease, box-shadow 0.2s ease;
}

/* Bar colors for different operations - now using theme variables */
.bar-display {
  --bar-default-color: var(--color-foreground-muted);
  --bar-compare-color: var(--color-info);
  --bar-swap-color: var(--color-error);
  --bar-select-color: var(--color-warning);
  --bar-insert-color: var(--color-success);
  --bar-merge-color: var(--color-accent);
  --bar-sorted-color: var(--color-success);
  --bar-active-color: var(--color-info);
  --bar-border-color: var(--color-foreground);
  /* Additional step types for counting/bucketed sorts */
  --bar-count-color: var(--color-warning);
  --bar-collect-color: var(--color-accent);
  --bar-write-color: var(--color-success);
  /* Advanced notes */
  --bar-note-color: var(--color-foreground-subtle);
}

/* Step indicator */
.step-indicator {
  position: absolute;
  top: 1rem;
  left: 50%;
  transform: translateX(-50%);
  background: var(--color-panel-background);
  border: 2px solid var(--color-accent);
  border-radius: 8px;
  padding: 0.75rem 1rem;
  box-shadow: 0 4px 6px var(--color-shadow);
  z-index: 10;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.25rem;
  min-width: 200px;
}

.step-type {
  font-size: 0.75rem;
  font-weight: 700;
  color: var(--color-accent);
  text-transform: uppercase;
  letter-spacing: 0.05em;
}

.step-description {
  font-size: 0.875rem;
  color: var(--color-foreground);
  text-align: center;
  line-height: 1.3;
}

/* Array info */
.array-info {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.75rem 1rem;
  background: var(--color-panel-background);
  border-top: 1px solid var(--color-border);
  font-size: 0.875rem;
  color: var(--color-foreground-muted);
}

.array-size,
.array-range {
  font-weight: 500;
}

/* Empty state */
.bar-display.empty {
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 300px;
}

.empty-message {
  text-align: center;
  color: var(--color-foreground-muted);
}

.empty-message p {
  margin: 0.25rem 0;
}

.empty-message p:first-child {
  font-size: 1.125rem;
  font-weight: 500;
}

.empty-message p:last-child {
  font-size: 0.875rem;
}

/* Responsive design */
@media (max-width: 767px) {
  .bar-display {
    min-height: 600px;
  }
  
  .bar-container {
    padding: 16px;
  }
  
  .step-indicator {
    position: relative;
    top: auto;
    left: auto;
    transform: none;
    margin-bottom: 1rem;
    width: 100%;
  }
  
  .array-info {
    flex-direction: column;
    gap: 0.5rem;
    text-align: center;
  }
}

/* Animation classes */
.bar.animate-compare {
  animation: pulse 0.5s ease-in-out;
}

.bar.animate-swap {
  animation: shake 0.3s ease-in-out;
}

.bar.animate-insert {
  animation: slideIn 0.4s ease-out;
}

.bar.animate-merge {
  animation: glow 0.6s ease-in-out;
}

/* Keyframe animations */
@keyframes pulse {
  0%, 100% { transform: scale(1); }
  50% { transform: scale(1.1); }
}

@keyframes shake {
  0%, 100% { transform: translateX(0); }
  25% { transform: translateX(-2px); }
  75% { transform: translateX(2px); }
}

@keyframes slideIn {
  0% { transform: translateY(-10px); opacity: 0; }
  100% { transform: translateY(0); opacity: 1; }
}

@keyframes glow {
  0%, 100% { box-shadow: 0 1px 3px var(--color-shadow); }
  50% { box-shadow: 0 0 20px var(--bar-merge-color); }
}

/* Reduced motion support */
@media (prefers-reduced-motion: reduce) {
  .bar {
    transition: none !important;
  }
  
  .bar:hover {
    transform: none;
  }
  
  .step-indicator {
    animation: none;
  }
}

/* Focus styles for accessibility */
.bar:focus {
  outline: 2px solid var(--color-accent);
  outline-offset: 2px;
}

/* Print styles */
@media print {
  .bar-display {
    break-inside: avoid;
  }
  
  .step-indicator {
    display: none;
  }
  
  .bar {
    border: 1px solid var(--color-foreground);
    box-shadow: none;
  }
}

/* Custom scrollbar for overflow */
.bar-container::-webkit-scrollbar {
  width: 8px;
  height: 8px;
}

.bar-container::-webkit-scrollbar-track {
  background: var(--color-panel-background);
  border-radius: 4px;
}

.bar-container::-webkit-scrollbar-thumb {
  background: var(--color-border);
  border-radius: 4px;
}

.bar-container::-webkit-scrollbar-thumb:hover {
  background: var(--color-accent);
}

/* Loading state */
.bar-display.loading .bar-container {
  opacity: 0.7;
}

.bar-display.loading::after {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 40px;
  height: 40px;
  border: 4px solid var(--color-border);
  border-top: 4px solid var(--color-accent);
  border-radius: 50%;
  animation: spin 1s linear infinite;
}

@keyframes spin {
  0% { transform: translate(-50%, -50%) rotate(0deg); }
  100% { transform: translate(-50%, -50%) rotate(360deg); }
}
