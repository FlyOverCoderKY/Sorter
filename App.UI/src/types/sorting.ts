// Sorting algorithm types
export type AlgorithmType =
  | 'bubble'
  | 'selection'
  | 'insertion'
  | 'merge'
  | 'quick'
  | 'heap'
  | 'shell'
  | 'comb'
  | 'cocktail'
  | 'counting'
  | 'naturalMerge'
  | 'intro'
  | 'timsort'
  | 'bitonic'
  | 'tree'
  | 'bogo'
  | 'bozo'
  | 'stooge'
  | 'gnome'
  | 'oddEven'
  | 'stableSelection'
  | 'radix'
  | 'bucket'
  | 'pigeonhole';

// State machine states
export type AppState = 'IDLE' | 'GENERATING' | 'SORTING' | 'PAUSED' | 'COMPLETED';

// Performance metrics
export interface PerformanceMetrics {
  steps: number;
  comparisons: number;
  swaps: number;
  executionTime: number;
  memoryUsage: number;
}

// Sorting step for visualization
export interface SortingStep {
  type:
    | 'compare'
    | 'swap'
    | 'merge'
    | 'select'
    | 'insert'
    | 'count'
    | 'collect'
    | 'write'
    | 'bucket'
    | 'partition'
    | 'heapify'
    | 'gap'
    | 'note';
  indices: number[];
  values: number[];
  description: string;
  timestamp: number;
}

// Web Worker message types
export interface WorkerMessage {
  type: 'SORT_REQUEST' | 'PAUSE_REQUEST' | 'RESUME_REQUEST' | 'CANCEL_REQUEST';
  algorithm: AlgorithmType;
  array: number[];
  speed: number;
  id: string;
}

export interface WorkerResponse {
  type: 'SORT_STEP' | 'SORT_COMPLETE' | 'SORT_PAUSED' | 'SORT_CANCELLED' | 'ERROR';
  step?: SortingStep;
  metrics?: PerformanceMetrics;
  sortedArray?: number[];
  currentArray?: number[];
  error?: string;
  id: string;
}

// Algorithm configuration
export interface AlgorithmConfig {
  name: string;
  type: AlgorithmType;
  timeComplexity: {
    best: string;
    average: string;
    worst: string;
  };
  spaceComplexity: string;
  description: string;
  pseudoCode: string[];
}

// App configuration
export interface AppConfig {
  defaultArraySize: number;
  minArraySize: number;
  maxArraySize: number;
  defaultSpeed: number;
  minSpeed: number;
  maxSpeed: number;
  animationDuration: number;
}
