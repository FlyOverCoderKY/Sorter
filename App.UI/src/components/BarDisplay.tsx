import React, { useRef, useEffect, useMemo, useCallback, useState } from 'react';
import { SortingStep } from '../types/sorting';
import { calculateBarDimensions } from '../utils/arrayUtils';
import './BarDisplay.css';

interface BarDisplayProps {
  array: number[];
  currentStep: SortingStep | null;
  isRunning: boolean;
}

// Memoized individual bar component to prevent unnecessary re-renders
const Bar = React.memo<{
  value: number;
  index: number;
  maxValue: number;
  barWidth: number;
  spacing: number;
  isLast: boolean;
  currentStep: SortingStep | null;
  isRunning: boolean;
  isSizeChanging: boolean;
}>(({ value, index, maxValue, barWidth, spacing, isLast, currentStep, isRunning, isSizeChanging }) => {
  // Memoized color calculation
  const backgroundColor = useMemo(() => {
    if (!currentStep || !isRunning) {
      return 'var(--bar-default-color)';
    }

    const { type, indices } = currentStep;

    // Highlight bars involved in current operation
    if (indices.includes(index)) {
      switch (type) {
        case 'compare':
          return 'var(--bar-compare-color)';
        case 'swap':
          return 'var(--bar-swap-color)';
        case 'select':
          return 'var(--bar-select-color)';
        case 'insert':
          return 'var(--bar-insert-color)';
        case 'merge':
          return 'var(--bar-merge-color)';
        default:
          return 'var(--bar-active-color)';
      }
    }

    // Highlight sorted portions (for algorithms like insertion sort)
    if (type === 'insert' && index < indices[0]) {
      return 'var(--bar-sorted-color)';
    }

    return 'var(--bar-default-color)';
  }, [currentStep, isRunning, index]);

  // Memoized border calculation
  const border = useMemo(() => {
    if (!currentStep || !isRunning) return 'none';
    
    const { indices } = currentStep;
    
    if (indices.includes(index)) {
      return '2px solid var(--bar-border-color)';
    }
    
    return 'none';
  }, [currentStep, isRunning, index]);

  // Memoized tooltip content
  const tooltip = useMemo(() => {
    let tooltipText = `Position ${index + 1}, Value: ${value}`;
    
    if (currentStep && isRunning) {
      const { indices, description } = currentStep;
      if (indices.includes(index)) {
        tooltipText += ` - ${description}`;
      }
    }
    
    return tooltipText;
  }, [currentStep, isRunning, index, value]);

  // Only apply transitions when size is changing, not during sorting
  const transition = isSizeChanging ? 'width 0.3s ease, margin-right 0.3s ease' : 'none';

  return (
    <div
      className="bar"
      style={{
        height: `${(value / maxValue) * 100}%`,
        width: `${barWidth}px`,
        marginRight: isLast ? '0px' : `${spacing}px`,
        backgroundColor,
        border,
        opacity: 1,
        transition
      }}
      role="img"
      aria-label={tooltip}
      title={tooltip}
      data-index={index}
      data-value={value}
    />
  );
});

Bar.displayName = 'Bar';

const BarDisplay = React.memo<BarDisplayProps>(({ array, currentStep, isRunning }) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const maxValue = Math.max(...array, 1);
  const [isSizeChanging, setIsSizeChanging] = useState(false);
  const [prevArraySize, setPrevArraySize] = useState(array.length);

  // Calculate bar dimensions based on container width and array size
  const barDimensions = useMemo(() => {
    if (!containerRef.current) return { barWidth: 4, spacing: 2, totalWidth: 0 };
    
    const containerWidth = containerRef.current.clientWidth;
    return calculateBarDimensions(containerWidth, array.length);
  }, [array.length]);

  // Track array size changes for smooth transitions
  useEffect(() => {
    if (prevArraySize !== array.length) {
      setIsSizeChanging(true);
      setPrevArraySize(array.length);
      
      // Reset the size changing flag after transition completes
      const timer = setTimeout(() => {
        setIsSizeChanging(false);
      }, 300); // Match the transition duration
      
      return () => clearTimeout(timer);
    }
  }, [array.length, prevArraySize]);

  // Update bar dimensions on window resize
  useEffect(() => {
    const handleResize = () => {
      // Force re-render by updating a state or ref
      if (containerRef.current) {
        containerRef.current.style.width = containerRef.current.clientWidth + 'px';
      }
    };

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  if (array.length === 0) {
    return (
      <div className="bar-display empty" ref={containerRef}>
        <div className="flex flex-col items-center gap-2 empty-message">
          <span className="text-lg">No data to display</span>
          <span className="text-sm text-gray-500">Generate an array to get started</span>
        </div>
      </div>
    );
  }

  return (
    <div className="bar-display" ref={containerRef}>
      <div 
        className="bar-container"
        style={{
          maxWidth: '100%'
        }}
      >
        {array.map((value, index) => (
          <Bar
            key={index}
            value={value}
            index={index}
            maxValue={maxValue}
            barWidth={barDimensions.barWidth}
            spacing={barDimensions.spacing}
            isLast={index === array.length - 1}
            currentStep={currentStep}
            isRunning={isRunning}
            isSizeChanging={isSizeChanging}
          />
        ))}
      </div>
      
      {/* Current step indicator */}
      {currentStep && isRunning && (
        <div className="flex flex-col gap-1 step-indicator" role="status" aria-live="polite">
          <span className="font-semibold text-sm">{currentStep.type.toUpperCase()}</span>
          <span className="text-sm">{currentStep.description}</span>
        </div>
      )}
      
      {/* Array info */}
      <div className="flex gap-4 array-info">
        <span className="text-sm">Array Size: {array.length}</span>
        <span className="text-sm">Range: {Math.min(...array)} - {Math.max(...array)}</span>
      </div>
    </div>
  );
});

BarDisplay.displayName = 'BarDisplay';

export default BarDisplay;
