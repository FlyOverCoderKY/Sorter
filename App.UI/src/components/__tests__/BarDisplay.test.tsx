import React from 'react';
import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import BarDisplay from '../BarDisplay';
import { SortingStep } from '../../types/sorting';

const baseArray = [10, 20, 30, 40, 25, 15];

function makeStep(type: SortingStep['type'], indices: number[], values: number[], description: string): SortingStep {
  return { type, indices, values, description, timestamp: 0 };
}

describe('BarDisplay snapshots', () => {
  it('renders idle state', () => {
    const { container } = render(
      <BarDisplay array={baseArray} currentStep={null} isRunning={false} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders compare state', () => {
    const step = makeStep('compare', [1, 2], [baseArray[1], baseArray[2]], 'Comparing');
    const { container } = render(
      <BarDisplay array={baseArray} currentStep={step} isRunning={true} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders swap state', () => {
    const step = makeStep('swap', [2, 3], [baseArray[2], baseArray[3]], 'Swapping');
    const { container } = render(
      <BarDisplay array={baseArray} currentStep={step} isRunning={true} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders insert state', () => {
    const step = makeStep('insert', [1], [baseArray[1]], 'Inserting');
    const { container } = render(
      <BarDisplay array={baseArray} currentStep={step} isRunning={true} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders merge state', () => {
    const step = makeStep('merge', [0, 1], [baseArray[0], baseArray[1]], 'Merging');
    const { container } = render(
      <BarDisplay array={baseArray} currentStep={step} isRunning={true} />
    );
    expect(container).toMatchSnapshot();
  });
});


