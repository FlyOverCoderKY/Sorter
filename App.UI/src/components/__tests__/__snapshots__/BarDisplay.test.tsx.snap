// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarDisplay snapshots > renders compare state 1`] = `
<div>
  <div
    class="bar-display"
  >
    <div
      class="bar-container"
      style="max-width: 100%;"
    >
      <div
        aria-label="Position 1, Value: 10"
        class="bar"
        data-index="0"
        data-value="10"
        role="img"
        style="height: 25%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 1, Value: 10"
      />
      <div
        aria-label="Position 2, Value: 20 - Comparing"
        class="bar"
        data-index="1"
        data-value="20"
        role="img"
        style="height: 50%; width: 4px; margin-right: 2px; background-color: var(--bar-compare-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 2, Value: 20 - Comparing"
      />
      <div
        aria-label="Position 3, Value: 30 - Comparing"
        class="bar"
        data-index="2"
        data-value="30"
        role="img"
        style="height: 75%; width: 4px; margin-right: 2px; background-color: var(--bar-compare-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 3, Value: 30 - Comparing"
      />
      <div
        aria-label="Position 4, Value: 40"
        class="bar"
        data-index="3"
        data-value="40"
        role="img"
        style="height: 100%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 4, Value: 40"
      />
      <div
        aria-label="Position 5, Value: 25"
        class="bar"
        data-index="4"
        data-value="25"
        role="img"
        style="height: 62.5%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 5, Value: 25"
      />
      <div
        aria-label="Position 6, Value: 15"
        class="bar"
        data-index="5"
        data-value="15"
        role="img"
        style="height: 37.5%; width: 4px; margin-right: 0px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 6, Value: 15"
      />
    </div>
    <div
      aria-live="polite"
      class="flex flex-col gap-1 step-indicator"
      role="status"
    >
      <span
        class="font-semibold text-sm"
      >
        COMPARE
      </span>
      <span
        class="text-sm"
      >
        Comparing
      </span>
    </div>
    <div
      class="flex gap-4 array-info"
    >
      <span
        class="text-sm"
      >
        Array Size: 
        6
      </span>
      <span
        class="text-sm"
      >
        Range: 
        10
         - 
        40
      </span>
    </div>
  </div>
</div>
`;

exports[`BarDisplay snapshots > renders idle state 1`] = `
<div>
  <div
    class="bar-display"
  >
    <div
      class="bar-container"
      style="max-width: 100%;"
    >
      <div
        aria-label="Position 1, Value: 10"
        class="bar"
        data-index="0"
        data-value="10"
        role="img"
        style="height: 25%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 1, Value: 10"
      />
      <div
        aria-label="Position 2, Value: 20"
        class="bar"
        data-index="1"
        data-value="20"
        role="img"
        style="height: 50%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 2, Value: 20"
      />
      <div
        aria-label="Position 3, Value: 30"
        class="bar"
        data-index="2"
        data-value="30"
        role="img"
        style="height: 75%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 3, Value: 30"
      />
      <div
        aria-label="Position 4, Value: 40"
        class="bar"
        data-index="3"
        data-value="40"
        role="img"
        style="height: 100%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 4, Value: 40"
      />
      <div
        aria-label="Position 5, Value: 25"
        class="bar"
        data-index="4"
        data-value="25"
        role="img"
        style="height: 62.5%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 5, Value: 25"
      />
      <div
        aria-label="Position 6, Value: 15"
        class="bar"
        data-index="5"
        data-value="15"
        role="img"
        style="height: 37.5%; width: 4px; margin-right: 0px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 6, Value: 15"
      />
    </div>
    <div
      class="flex gap-4 array-info"
    >
      <span
        class="text-sm"
      >
        Array Size: 
        6
      </span>
      <span
        class="text-sm"
      >
        Range: 
        10
         - 
        40
      </span>
    </div>
  </div>
</div>
`;

exports[`BarDisplay snapshots > renders insert state 1`] = `
<div>
  <div
    class="bar-display"
  >
    <div
      class="bar-container"
      style="max-width: 100%;"
    >
      <div
        aria-label="Position 1, Value: 10"
        class="bar"
        data-index="0"
        data-value="10"
        role="img"
        style="height: 25%; width: 4px; margin-right: 2px; background-color: var(--bar-sorted-color); opacity: 1; transition: none;"
        title="Position 1, Value: 10"
      />
      <div
        aria-label="Position 2, Value: 20 - Inserting"
        class="bar"
        data-index="1"
        data-value="20"
        role="img"
        style="height: 50%; width: 4px; margin-right: 2px; background-color: var(--bar-insert-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 2, Value: 20 - Inserting"
      />
      <div
        aria-label="Position 3, Value: 30"
        class="bar"
        data-index="2"
        data-value="30"
        role="img"
        style="height: 75%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 3, Value: 30"
      />
      <div
        aria-label="Position 4, Value: 40"
        class="bar"
        data-index="3"
        data-value="40"
        role="img"
        style="height: 100%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 4, Value: 40"
      />
      <div
        aria-label="Position 5, Value: 25"
        class="bar"
        data-index="4"
        data-value="25"
        role="img"
        style="height: 62.5%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 5, Value: 25"
      />
      <div
        aria-label="Position 6, Value: 15"
        class="bar"
        data-index="5"
        data-value="15"
        role="img"
        style="height: 37.5%; width: 4px; margin-right: 0px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 6, Value: 15"
      />
    </div>
    <div
      aria-live="polite"
      class="flex flex-col gap-1 step-indicator"
      role="status"
    >
      <span
        class="font-semibold text-sm"
      >
        INSERT
      </span>
      <span
        class="text-sm"
      >
        Inserting
      </span>
    </div>
    <div
      class="flex gap-4 array-info"
    >
      <span
        class="text-sm"
      >
        Array Size: 
        6
      </span>
      <span
        class="text-sm"
      >
        Range: 
        10
         - 
        40
      </span>
    </div>
  </div>
</div>
`;

exports[`BarDisplay snapshots > renders merge state 1`] = `
<div>
  <div
    class="bar-display"
  >
    <div
      class="bar-container"
      style="max-width: 100%;"
    >
      <div
        aria-label="Position 1, Value: 10 - Merging"
        class="bar"
        data-index="0"
        data-value="10"
        role="img"
        style="height: 25%; width: 4px; margin-right: 2px; background-color: var(--bar-merge-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 1, Value: 10 - Merging"
      />
      <div
        aria-label="Position 2, Value: 20 - Merging"
        class="bar"
        data-index="1"
        data-value="20"
        role="img"
        style="height: 50%; width: 4px; margin-right: 2px; background-color: var(--bar-merge-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 2, Value: 20 - Merging"
      />
      <div
        aria-label="Position 3, Value: 30"
        class="bar"
        data-index="2"
        data-value="30"
        role="img"
        style="height: 75%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 3, Value: 30"
      />
      <div
        aria-label="Position 4, Value: 40"
        class="bar"
        data-index="3"
        data-value="40"
        role="img"
        style="height: 100%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 4, Value: 40"
      />
      <div
        aria-label="Position 5, Value: 25"
        class="bar"
        data-index="4"
        data-value="25"
        role="img"
        style="height: 62.5%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 5, Value: 25"
      />
      <div
        aria-label="Position 6, Value: 15"
        class="bar"
        data-index="5"
        data-value="15"
        role="img"
        style="height: 37.5%; width: 4px; margin-right: 0px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 6, Value: 15"
      />
    </div>
    <div
      aria-live="polite"
      class="flex flex-col gap-1 step-indicator"
      role="status"
    >
      <span
        class="font-semibold text-sm"
      >
        MERGE
      </span>
      <span
        class="text-sm"
      >
        Merging
      </span>
    </div>
    <div
      class="flex gap-4 array-info"
    >
      <span
        class="text-sm"
      >
        Array Size: 
        6
      </span>
      <span
        class="text-sm"
      >
        Range: 
        10
         - 
        40
      </span>
    </div>
  </div>
</div>
`;

exports[`BarDisplay snapshots > renders swap state 1`] = `
<div>
  <div
    class="bar-display"
  >
    <div
      class="bar-container"
      style="max-width: 100%;"
    >
      <div
        aria-label="Position 1, Value: 10"
        class="bar"
        data-index="0"
        data-value="10"
        role="img"
        style="height: 25%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 1, Value: 10"
      />
      <div
        aria-label="Position 2, Value: 20"
        class="bar"
        data-index="1"
        data-value="20"
        role="img"
        style="height: 50%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 2, Value: 20"
      />
      <div
        aria-label="Position 3, Value: 30 - Swapping"
        class="bar"
        data-index="2"
        data-value="30"
        role="img"
        style="height: 75%; width: 4px; margin-right: 2px; background-color: var(--bar-swap-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 3, Value: 30 - Swapping"
      />
      <div
        aria-label="Position 4, Value: 40 - Swapping"
        class="bar"
        data-index="3"
        data-value="40"
        role="img"
        style="height: 100%; width: 4px; margin-right: 2px; background-color: var(--bar-swap-color); border: 2px solid var(--bar-border-color); opacity: 1; transition: none;"
        title="Position 4, Value: 40 - Swapping"
      />
      <div
        aria-label="Position 5, Value: 25"
        class="bar"
        data-index="4"
        data-value="25"
        role="img"
        style="height: 62.5%; width: 4px; margin-right: 2px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 5, Value: 25"
      />
      <div
        aria-label="Position 6, Value: 15"
        class="bar"
        data-index="5"
        data-value="15"
        role="img"
        style="height: 37.5%; width: 4px; margin-right: 0px; background-color: var(--bar-default-color); opacity: 1; transition: none;"
        title="Position 6, Value: 15"
      />
    </div>
    <div
      aria-live="polite"
      class="flex flex-col gap-1 step-indicator"
      role="status"
    >
      <span
        class="font-semibold text-sm"
      >
        SWAP
      </span>
      <span
        class="text-sm"
      >
        Swapping
      </span>
    </div>
    <div
      class="flex gap-4 array-info"
    >
      <span
        class="text-sm"
      >
        Array Size: 
        6
      </span>
      <span
        class="text-sm"
      >
        Range: 
        10
         - 
        40
      </span>
    </div>
  </div>
</div>
`;
